src/hardware/input/mouseif_ps2_bios.cpp

```diff
@@ -359,148 +697,109 @@ bool MOUSEPS2_NotifyButton(const MouseButtons12S new_buttons_12S,
-// Other solution might be to put interrupt lines low ion BIOS implementation,
-// like this is done in DOSBox X.
+enum class BiosRetVal : uint8_t {
+	Success         = 0x00,
+	InvalidFunction = 0x01,
+	InvalidInput    = 0x02,
+	InterfaceError  = 0x03,
+	NeedToResend    = 0x04,
+	NoDeviceHandler = 0x05,
+};
 
-static bool packet_4bytes = false;
+static bool bios_is_flushing   = false;
+static uint8_t bios_frame_size = 3;
 
 static bool callback_init    = false;
 static uint16_t callback_seg = 0;
 static uint16_t callback_ofs = 0;
 static RealPt ps2_callback   = 0;
 
-void MOUSEBIOS_Reset()
-{
-	cmd_reset();
-	PIC_SetIRQMask(mouse_predefined.IRQ_PS2, false); // lower IRQ line
-	MOUSEVMM_Deactivate(); // VBADOS seems to expect this
-}
+std::vector<uint8_t> bios_buffer = {};
 
-void MOUSEBIOS_SetCallback(const uint16_t pseg, const uint16_t pofs)
-{
-	if ((pseg == 0) && (pofs == 0)) {
-		callback_init = false;
-	} else {
-		callback_init = true;
-		callback_seg  = pseg;
-		callback_ofs  = pofs;
-	}
-}
-
-bool MOUSEBIOS_SetPacketSize(const uint8_t packet_size)
-{
-	if (packet_size == 3)
-		packet_4bytes = false;
-	else if (packet_size == 4)
-		packet_4bytes = true;
-	else
-		return false; // unsupported packet size
-
-	return true;
-}
-
-bool MOUSEBIOS_SetSampleRate(const uint8_t rate_id)
-{
-	if (rate_id >= list_rates_hz.size())
-		return false;
-
-	cmd_set_sample_rate(list_rates_hz[rate_id]);
-	return true;
-}
-
-bool MOUSEBIOS_SetResolution(const uint8_t res_id)
-{
-	if (res_id >= list_resolutions.size())
-		return false;
-
-	cmd_set_sample_rate(list_resolutions[res_id]);
-	return true;
-}
-
-void MOUSEBIOS_SetScaling21(const bool enable)
-{
-	cmd_set_scaling_21(enable);
-}
-
-bool MOUSEBIOS_Enable()
+static bool bios_enable()
 {
 	mouse_shared.active_bios = callback_init;
+	bios_buffer.clear();
 	MOUSE_UpdateGFX();
+	cmd_set_reporting(true);
 	return callback_init;
 }
 
-bool MOUSEBIOS_Disable()
+static bool bios_disable()
 {
 	mouse_shared.active_bios = false;
+	bios_buffer.clear();
 	MOUSE_UpdateGFX();
+	cmd_set_reporting(false);
 	return true;
 }
 
-uint8_t MOUSEBIOS_GetResolution()
+static bool bios_is_aux_byte_waiting()
 {
-	return counts_mm;
+	const auto byte = IO_ReadB(0x64);
+	return bit::is(byte, b0) && bit::is(byte, b5);
 }
 
-uint8_t MOUSEBIOS_GetSampleRate()
+static void bios_flush_aux()
 {
-	return rate_hz;
-}
+	constexpr uint8_t max_wait_ms = 30;
 
-uint8_t MOUSEBIOS_GetStatus()
-{
-	union {
-		uint8_t data = 0;
-
-		bit_view<0, 1> left;
-		bit_view<1, 1> right;
-		bit_view<2, 1> middle;
-		// bit 3 - reserved
-		bit_view<4, 1> scaling_21;
-		bit_view<5, 1> reporting;
-		bit_view<6, 1> mode_remote;
-		// bit 7 - reserved
-	} ret;
+	bios_is_flushing = true;
+	bios_buffer.clear();
 
-	ret.left   = buttons.left;
-	ret.right  = buttons.right;
-	ret.middle = buttons.middle;
+	bool has_more = true;
+	while (has_more) {
+		while (bios_is_aux_byte_waiting()) {
+			IO_ReadB(0x60);
```

Comment:
Would it be possible to extract such port names to meaningfully named constants?

Reply:
Done - see last commit. But please keep in mind, that there are many more I/O ports emulated throughout the whole emulator, and they probably almost never use defined constants.
Also, I'm a bit afraid someone might get confused when analyzing the code with some hardware reference, as there are no standard port names in the PC world.

@kcgen What's your opinion?

Reply:
Well, I don't see how it can be more confusing than what we currently have because you can always simply check  what the constant points to. Yes, it's bothering me that most of the codebase doesn't use symbolic constant for ports; you need to dig out a reference manual to make any sense of it. I would only remember all the port hex codes if I was doing this day in, day out (and probably not even then).

The names seem pretty well defined to me, check this:
https://bochs.sourceforge.io/techspec/PORTS.LST

Mt experience is the opposite; when I was coding in assembly in the 90s, I pretty much had to use the reference docs constantly because I couldn't remember 90% of all the hex codes.

E.g. how much easier would be to read the GUS code if all these ports were referenced by their symbolic names:

```
0240-024F ----	Gravis Ultra Sound by Advanced Gravis
	 The I/O address range is dipswitch selectable from:
	   0200-020F and 0300-030F
	   0210-021F and 0310-031F
	   0220-022F and 0320-032F
	   0230-023F and 0330-033F
	   0240-024F and 0340-034F
	   0250-025F and 0350-035F
	   0260-026F and 0360-036F
	   0270-027F and 0370-037F

0240	w	Mix Control register
		 bit 6	   Control Register Select (see 024B)
		 bit 5	   Enable MIDI Loopback
		 bit 4	   Combine GF1 IRQ with MIDI IRQ
		 bit 3	   Enable Latches
		 bit 2	   Enable MIC IN
		 bit 1	   Disable LINE OUT
		 bit 0	   Disable LINE IN
0241	r	Read Data
	w	Trigger Timer
0246	r	IRQ Status Register
		 bit 7	   DMA TC IRQ
		 bit 6	   Volume Ramp IRQ
		 bit 5	   WaveTable IRQ
		 bit 3	   Timer 2 IRQ
		 bit 2	   Timer 1 IRQ
		 bit 1	   MIDI Receive IRQ
		 bit 0	   MIDI Transmit IRQ
0248	r/w	Timer Control Reg
		 Same a ADLIB Board (look at 0200)
0249	w	Timer Data
		 bit 7	   Reset Timr IRQ
		 bit 6	   Mask Timer 1
		 bit 5	   Mask Timer 2
		 bit 1	   Timer 2 Start
		 bit 0	   Timer 1 Start
024B	w	IRQ Control Register (0240 bit 6 = 1)
		 bit 6	   Combine Both IRQ
		 bit 5-3   MIDI IRQ Selector
			   = 000  No IRQ
			   = 001  IRQ 2
			   = 010  IRQ 5
			   = 011  IRQ 3
			   = 100  IRQ 7
			   = 101  IRQ 11
			   = 110  IRQ 12
			   = 111  IRQ 15
		 bit 2-0   GF1 IRQ Selector
			   = 000  No IRQ
			   = 001  IRQ 2
			   = 010  IRQ 5
			   = 011  IRQ 3
			   = 100  IRQ 7
			   = 101  IRQ 11
			   = 110  IRQ 12
			   = 111  IRQ 15
		DMA Control Register (0240 bit 6 = 0)
		 bit 6	   Combine Both IRQ
		 bit 5-3   DMA Select Register 2
			   = 000  No DMA
			   = 001  DMA 1
			   = 010  DMA 3
			   = 011  DMA 5
			   = 100  DMA 6
			   = 101  DMA 7 
		 bit 2-0   DMA Select Register 1
			   = 000  No DMA
			   = 001  DMA 1
			   = 010  DMA 3
			   = 011  DMA 5
			   = 100  DMA 6
			   = 101  DMA 7 
024F	r/w	Register Controls (rev 3.4+)
```

Reply:
> Yes, it's bothering me that most of the codebase doesn't use symbolic constant for ports; you need to dig out a reference manual to make any sense of it. 

Yeah, it sure would be nice to have a block of `constexpr` self-documenting ports named after what they do. 

@FeralChild64 - all these named bits in this PR make it so much easier to read than having to visually chew through hard coded masks and shifts everywhere. The code reads very naturally :+1:

Reply:
@johnnovak By _standard names_ I mean something like visible here: https://github.com/MEGA65/open-roms/blob/master/src/aliases/%2Caliases_hardware.s. For example, `EXTCOL` is a "canonical" name of $D020 register, and it came from the official documentation. AFAIK there is nothing like that for PC-compatible I/O ports, everyone invents own names.

Reply:
I like the naming standard you've started @FeralChild64 as it should be expandable for all the devices.

`port_num_<device>_<action>`, with `<action>` itself being `read`, `write`, `status`, or possibly a more specifically named action if the port only performs single task.

Reply:
> @johnnovak By _standard names_ I mean something like visible here: https://github.com/MEGA65/open-roms/blob/master/src/aliases/%2Caliases_hardware.s. For example, `EXTCOL` is a "canonical" name of $D020 register, and it came from the official documentation. AFAIK there is nothing like that for PC-compatible I/O ports, everyone invents own names.

Ah ok I see. Well, I still think it's better to come up with a symbolic name based on existing documentation, books, etc., even if it's not standardised. How I see it, that does not really matter if things are called slighy differently in various docs, a status port will be called still a status port, and it's easier to make a connection between slightly different names than constantly having to look up random hex codes.